libsnark/zk_proof_systems/plonk/plonk.tcc

```diff
@@ -0,0 +1,412 @@
+/// equal to the number of gates. L_basis is the Lagrange basis.
+template<typename FieldT>
+void plonk_compute_selector_polynomials(
+    const std::vector<std::vector<FieldT>> &gates_matrix_transpose,
+    std::vector<polynomial<FieldT>> &Q_polys)
+{
+    assert(gates_matrix_transpose.size() == Q_polys.size());
+    assert(gates_matrix_transpose[0].size() == Q_polys[0].size());
+    size_t num_qpolys = gates_matrix_transpose.size();
+    for (size_t i = 0; i < num_qpolys; ++i) {
+        std::vector<FieldT> q_vec = gates_matrix_transpose[i];
+        plonk_interpolate_polynomial_from_points<FieldT>(q_vec, Q_polys[i]);
+    }
+};
+
+template<typename FieldT>
+void plonk_compute_public_input_polynomial(
+    const std::vector<FieldT> &PI_points, polynomial<FieldT> &PI_poly)
+{
+    plonk_interpolate_polynomial_from_points<FieldT>(PI_points, PI_poly);
+};
+
+/// output: omega[0] are the n roots of unity, omega[1] are
+/// omega[0]*k1, omega[2] are omega[0]*k2; k1 H is a coset of H with
+/// generator omega_k1 distinct from H; k2 H is a coset of H with
+/// generator omega_k2, distinct from H and k1 H
+template<typename FieldT>
+void plonk_compute_roots_of_unity_omega(
+    const FieldT k1, const FieldT k2, std::vector<std::vector<FieldT>> &omega)
+{
+    assert(omega.size() == NUM_HSETS);
+    assert(omega[0].size() > 0);
+    size_t num_gates = omega[0].size();
+    // Get the n-th root of unity omega in Fq (n=8 is the number of
+    // constraints in the example). omega is a generator of the
+    // multiplicative subgroup H.  Example (2**32)-th primitive root
+    // of unity in the base field Fq of bls12-381 i.e. such that
+    // omega_base**(2**32) = 1. The bls12-381 prime q is such that any
+    // power of 2 divides (q-1). In particular 2**32|(q-1) and so the
+    // 2**32-th root of unity exists.
+    FieldT omega_base = libff::get_root_of_unity<FieldT>(num_gates);
+    for (size_t i = 0; i < num_gates; ++i) {
+        FieldT omega_i = libff::power(omega_base, libff::bigint<1>(i));
+        omega[base][i] = omega_i;
+        FieldT omega_k1_i = omega[base][i] * k1;
+        FieldT omega_k2_i = omega[base][i] * k2;
+        omega[base_k1][i] = omega_k1_i;
+        omega[base_k2][i] = omega_k2_i;
+    }
+}
+
+/// copy the roots of unity omega[base], omega[k1] and omega[k2] into
+/// a single vector H_gen representing the multiplicative subgroups
+/// H, k1 H and k2 H (see [GWC19] Sect. 8); \see
+/// plonk_compute_roots_of_unity_omega
+template<typename FieldT>
+void plonk_roots_of_unity_omega_to_subgroup_H(
+    const std::vector<std::vector<FieldT>> &omega, std::vector<FieldT> &H_gen)
+{
+    assert(omega.size() == NUM_HSETS);
+    assert(omega[0].size() > 0);
+    std::copy(omega[base].begin(), omega[base].end(), back_inserter(H_gen));
+    std::copy(
+        omega[base_k1].begin(), omega[base_k1].end(), back_inserter(H_gen));
+    std::copy(
+        omega[base_k2].begin(), omega[base_k2].end(), back_inserter(H_gen));
```

Comment:
Small point, but [erhaps we can just generate directly into the final `H_gen` array and avoid some copies.  We could potentially collect a few of these tiny points in an issue and address them later, rather than in this PR.

Reply:
Acknowledged. Created a separate issue https://github.com/clearmatics/libsnark/issues/51 containing a list of suggested optimisations for Plonk